File: ios/BleNative/Models/Buffer.swift
import Foundation

enum BufferType {
    case state
    case scan
    case valueChange
    case disconnect
    case stateRestore
    case name
} 

class Buffer {
    private(set) var items: [Any] = []
    let id: Int32
    let type: BufferType
    let relatedIdentifier: Int32?
    private(set) var invalidated: Bool = false
    private(set) var invalidatedReason: BleError? = nil
    
    init(id: Int32, type: BufferType, relatedIdentifier: Int32? = nil) {
        self.id = id
        self.type = type
        self.relatedIdentifier = relatedIdentifier
    }
    
    func invalidate(reason: BleError) {
        invalidated = true
        invalidatedReason = reason
    }
    
    func append(_ element: Any) {
        items.append(element)
    }
    
    func take(amount: Int? = nil, reversed: Bool = false, fulfill: Bool = true) -> [Any]? {
        return retrieve(andRemove: true, amount: amount, reversed: reversed, fulfill: fulfill)
    }
    
    func peek(amount: Int? = nil, reversed: Bool = false, fulfill: Bool = true) -> [Any]? {
        return retrieve(andRemove: false, amount: amount, reversed: reversed, fulfill: fulfill)
    }    
    
    private func retrieve(andRemove remove: Bool, amount: Int? = nil, reversed: Bool = false, fulfill: Bool = true) -> [Any]? {
        var amount = amount ?? items.count
        if fulfill && (amount > items.count || amount < 1) {
            return nil
        }
        if !fulfill {
            amount = max(min(items.count, amount), 0) 
        }
        
        let result: [Any]
        if reversed {
            result = Array(items.suffix(amount))
            if remove {
                items.removeLast(amount)
            }
        } else {
            result = Array(items.prefix(upTo: amount))
            if remove {
                items.removeFirst(amount)
            }
        }
        return result
    }
    
    func removeAll() {
        items.removeAll()
    }
}

File: android/src/main/java/com/polidea/blenative/models/Buffer.kt
package com.polidea.blenative.models

import kotlin.math.max
import kotlin.math.min

enum class BufferType {
    STATE,
    SCAN,
    VALUE_CHANGE,
    DISCONNECT,
    STATE_RESTORE,
    NAME,
    MTU
}

class Buffer(val id: Int, val type: BufferType, val relatedIdentifier: Int? = null) {

    private var items = mutableListOf<Any>()

    var invalidated = false
        private set

    var invalidatedReason: BleError? = null
        private set

    fun invalidate(reason: BleError) {
        invalidated = true
        invalidatedReason = reason
    }

    fun append(element: Any) {
        items.add(element)
    }

    fun appendAll(elements: List<Any>) {
        items.addAll(elements)
    }

    fun take(amount: Int? = null, lastItems: Boolean = false, fulfill: Boolean = true): List<Any>? {
        return retrieve(true, amount, lastItems, fulfill)
    }

    fun peek(amount: Int? = null, lastItems: Boolean = false, fulfill: Boolean = true): List<Any>? {
        return retrieve(false, amount, lastItems, fulfill)
    }

    private fun retrieve(remove: Boolean, amount: Int?, lastItems: Boolean, fulfill: Boolean): List<Any>? {
        var amount = amount ?: items.size
        if (fulfill && (amount > items.size || amount < 1)) {
            return null
        }
        if (!fulfill) {
            amount = max(min(items.size, amount), 0)
        }
        val result: List<Any>
        if (lastItems) {
            result = items.takeLast(amount)
            if (remove) {
                items = items.dropLast(amount).toMutableList()
            }
        } else {
            result = items.take(amount)
            if (remove) {
                items = items.drop(amount).toMutableList()
            }
        }
        return result
    }

    fun removeAll() {
        items.clear()
    }
}